ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/export/HeaderSpec.scala

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.export
+
+import java.io.{PipedInputStream, PipedOutputStream}
+
+import com.google.common.io.ByteStreams
+import org.scalatest.{Matchers, WordSpec}
+
+final class HeaderSpec extends WordSpec with Matchers {
+  "header" should {
+    "write and verify the version" in {
+      val input = new PipedInputStream
+      val output = new PipedOutputStream(input)
+      val header = new Header(version = "v1")
+
+      header.write(output)
+      output.close()
+
+      header.consumeAndVerify(input)
+      succeed
+    }
+
+    "throw if the preamble is wrong" in {
+      val input = new PipedInputStream
+      val output = new PipedOutputStream(input)
+      val header = new Header(version = "version two")
+
+      output.write("daml.kvutils.something\u0000version two\u0000".getBytes(Header.charset))
+      output.close()
+
+      a[Header.InvalidExportHeaderException] should be thrownBy header.consumeAndVerify(input)
+    }
+
+    "throw if the preamble is too long" in {
+      val input = new PipedInputStream
+      val output = new PipedOutputStream(input)
+      val header = new Header(version = "vee three")
+
+      output.write(s"${Header.preamble}.xyz\u0000version two\u0000".getBytes(Header.charset))
+      output.close()
+
+      a[Header.InvalidExportHeaderException] should be thrownBy header.consumeAndVerify(input)
+    }
+
+    "throw if the version is wrong" in {
+      val input = new PipedInputStream
+      val output = new PipedOutputStream(input)
+      val exportHeader = new Header(version = "wzn fuor")
+      val importHeader = new Header(version = "vsn four")
+
+      exportHeader.write(output)
+      output.close()
+
+      a[Header.InvalidExportHeaderException] should be thrownBy importHeader.consumeAndVerify(input)
+    }
+
+    "throw if the version is too long" in {
+      val input = new PipedInputStream
+      val output = new PipedOutputStream(input)
+      val exportHeader = new Header(version = "quersion 5.5")
+      val importHeader = new Header(version = "quersion 5")
+
+      exportHeader.write(output)
+      output.close()
+
+      a[Header.InvalidExportHeaderException] should be thrownBy importHeader.consumeAndVerify(input)
+    }
+
+    "flushes correctly, and does not buffer" in {
```

Comment:
This description is puzzling when reading the below test case as flushing is not apparent. I.e., the test case rather tells that `only the required part of the input is consumed by consumeAndVerify`.

Reply:
It tells us both, annoyingly. We can see that calling `output.write` after `header.write` doesn't end up with weird interleaving issues (which I had for a moment, which is why I wrote this test), and that calling `consumeAndVerify` does not buffer past the end of the header.

Reply:
With your explanation it indeed makes sense, thanks!